telethon/network/connection/connection.py

```diff
@@ -184,27 +189,71 @@ def _init_conn(self):
         data to Telegram to indicate which connection mode will
         be used.
         """
+        if self._codec.tag:
+            self._writer.write(self._codec.tag)
 
-    @abc.abstractmethod
     def _send(self, data):
-        """
-        This method should be implemented differently under each
-        connection mode and serialize the data into the packet
-        the way it should be sent through `self._writer`.
-        """
-        raise NotImplementedError
+        self._writer.write(self._codec.encode_packet(data))
 
-    @abc.abstractmethod
     async def _recv(self):
-        """
-        This method should be implemented differently under each
-        connection mode and deserialize the data from the packet
-        the way it should be read from `self._reader`.
-        """
-        raise NotImplementedError
+        return await self._codec.read_packet(self._reader)
 
     def __str__(self):
         return '{}:{}/{}'.format(
             self._ip, self._port,
             self.__class__.__name__.replace('Connection', '')
         )
+
+
+class ObfuscatedConnection(Connection):
+    """
+    Base class for "obfuscated" connections ("obfuscated2", "mtproto proxy")
+    """
+    """
+    This attribute should be redefined by subclasses
+    """
+    obfuscated_io = None
+
+    def _init_conn(self):
```

Comment:
Just double checking, but I guess it's not necessary to call `super()._init_conn()`?

Reply:
@Lonami I would say, it's important to NOT call it. Because obfuscated connection methods should have 64byte header as 1st packet and this header includes 4-byte indication (see obfuscate_tag attribute) of which packet codec will be used. While "plain" connections (except http and full) use 1st packet (1 or 4 byte) as an indication of which packet codec will be used (see tag attribute)